### PR TITLE
feat(sandbox): ergonomics — auth boundary, git cwd, editFile, persistence guidance

### PR DESCRIPTION
**Pull request by 2witstudios**
## Sandbox Ergonomics — auth boundary · git cwd · editFile

Five friction points an agent hit using the agent code-execution sandbox, captured from a real bug-fix run (it re-cloned a repo 3+ times, created divergent branches, lost track of its PR, and thrashed on blocked force-pushes). All changes stay behind the default-OFF `CODE_EXECUTION_ENABLED` kill-switch.

### What changed

1. **GitHub auth boundary made explicit.** `bash` has no GitHub credentials — only the dedicated `git_*`/`gh_*` tools carry OAuth. `command-policy` now redirects `gh …` / `git clone|fetch|pull|push` run in bash to those tools (UX redirect, ReDoS-safe linear regex; local git like `status`/`add`/`commit` still runs).
2. **Per-call `cwd` for git/gh tools.** `runGitInSandbox` already accepted `cwd` but the tools never passed it, pinning every op to `/workspace`. Now every repo-operating git/gh tool takes an optional `cwd` (validated, `path_escape` on escape), so the agent can operate on a repo cloned into a subdir.
3. **`editFile` tool.** Targeted string replacement (pure `applyEdit`, literal match, unique-unless-`replaceAll`) instead of whole-file `writeFile` rewrites or fragile `sed`.
4. **Workspace persistence made visible.** The sandbox filesystem *does* persist across turns (resume by session key + Fly hibernation, verified in `lifecycle.ts`/`session-key.ts`), but the model got no signal, so it re-cloned and re-opened PRs defensively. A short `SANDBOX_INSTRUCTIONS` block (gated on `codeExecutionEnabled`, threaded at the three chat prompt builders) tells it: `/workspace` persists — check `git_status`/`git_branch`/`gh_pr_list` before recreating; update the existing PR's branch.
5. **Tool-discovery overhead.** The same block names the key sandbox tools so the model skips a `tool_search` round-trip on the common path.

Plus **`git_push` force guard** (2c): `--force-with-lease` stays allowed on feature/PR branches but is refused for the default branch (`main`/`master`), and requires an explicit branch under `force`.

### Tests
TDD throughout (RED→GREEN). 282 lib sandbox tests + 88 web sandbox/system-prompt tests pass; full monorepo `typecheck` (incl. web build) and `lint` green. Covers: interceptor (allow/deny/chained/ReDoS), cwd validation + threading, force guard, `applyEdit` edge cases, `editSandboxFile` (escape/not-found/no-match/not-unique/oversize), `editFile` wiring, and `SANDBOX_INSTRUCTIONS` gating.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

### Review feedback addressed
- **P1 (force-push refspec bypass):** the force guard now parses the refspec *destination* (strips `+`, takes the segment after the last `:`, strips `refs/heads/`), so `HEAD:main` / `feature:refs/heads/master` can't slip past; `+`-prefixed refspecs count as forcing. Also refuses **deleting** the default branch via `:main`.
- **P2 (read-only exposure):** `editFile` — and the rest of the sandbox mutation surface (`bash`, `writeFile`, writing `git_*`/`gh_*`) — added to `WRITE_TOOLS`, so `filterToolsForReadOnly` strips them in read-only mode. Read-only sandbox tools (`readFile`, `git_status/diff/log`, `gh_*_list/view`) stay available.
- Fixed the CI `@pagespace/lib` typecheck failure (`WriteFileEntry.content` is `string | Uint8Array`).

### CODE REVIEW

File: apps/web/src/lib/ai/tools/sandbox-git-tools.ts [hunk 1]
```diff
@@ -334,28 +358,51 @@ export function createSandboxGitTools({ gitRunDeps, resolveContext, gate }: GitS
...
+      // can't see the sandbox's current branch from here.
+      if (force) {
+        if (!branch) {
+          return {
+            success: false as const,
+            error:
+              'Force-push requires an explicit branch so the target can be verified. Name the feature/PR branch to push to.',
+          };
+        }
+        if (DEFAULT_BRANCHES.has(branch.toLowerCase())) {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/tools/sandbox-git-tools.ts]: Fixed in 60e9e8840. The guard now parses the refspec **destination** rather than comparing the raw value: it strips a leading `+`, takes the segment after the last `:`, and strips a `refs/heads/` prefix before checking against `main`/`master`. So `HEAD:main` and `feature:refs/heads/master` are now refused, and a `+`-prefixed refspec (which forces per-refspec even without the `force` flag) is treated as forcing too. Added tests: `HEAD:main`, `feature:refs/heads/master`, and `+main` are rejected; `main:feature` (destination = feature) is allowed.

File: apps/web/src/lib/ai/tools/sandbox-tools.ts [hunk 2]
```diff
@@ -132,5 +141,16 @@ export function createSandboxTools({ runDeps, resolveContext, gate }: SandboxToo
         return readSandboxFile({ path, ctx: opened.ctx, deps: runDeps });
       },
     }),
+
+    editFile: tool({
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/tools/sandbox-tools.ts:145]: Fixed in 9951610ea. `editFile` is now in `WRITE_TOOLS` so `filterToolsForReadOnly` strips it in read-only mode. I extended this to the whole sandbox mutation surface, since the same gap applied: `bash` (can run arbitrary mutations), `writeFile`, and the writing git/gh tools (`git_clone/init/config/remote_add/add/reset/stash/commit/merge/rebase/checkout/branch/fetch/pull/push`, `gh_pr_create/merge/checkout`, `gh_issue_create`) are all now classified as writes. Read-only sandbox tools (`readFile`, `git_status/diff/log`, `gh_pr_list/view`, `gh_issue_list/view`) intentionally remain available. Added tool-filtering tests covering both partitions.

File: apps/web/src/lib/ai/tools/sandbox-git-tools.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/tools/sandbox-git-tools.ts]: Follow-up hardening in abbf020cb: while in the refspec-destination code I also closed the adjacent destructive variant — `git push origin :main` (empty-source refspec) *deletes* the remote default branch. The guard now refuses **force OR delete** to a `main`/`master` destination; normal fast-forward pushes and feature-branch deletes (`:feature`) stay allowed. Tests added for both.
